### DISCUSSION TITLE
Having different chat_histories per agent

### DISCUSSION THREAD

**Discussion by krlng** [microsoft/autogen#1303] [Q&A]
Currently, every agent has a complete list of all messages by all agents. I would like to have my agents store only certain messages: user_messages and own ones, but not those of the other agents. While I am pretty sure someone has asked something similar or there might even be an example notebook on how to do this, I was not able to find it. Maybe someone can point me into the right direction

**Comment by gagb** [comment]
Can you give a more concrete example? 

Also what's your use case. If an agent A is participating in conversation with B and C, you'd want A to store its state with respect to B and C?

**Reply by krlng** [reply]
Happy to give an example:

This are my agents (<name>: <description>)
* coder: has access to Neo4j and can answer analytical querries. Especially helpful if asked for trends, relationships and similiar
* press_observer: has full text access to articles of MediaX and MediaY, therefore can provide more context to answer questions that ask for reasons or detailed information. Not helpful to get trends.
* MediaX: can help with all questions regarding movies and actors as well as X and if user refers to "our" data.
* summarizer: Can create summaries of previous answers. Should talk as the last person, after all information to answer the question are already present in the conversation. Can not run functions!

Possible questions: 
* What are the most popular songs on Taylor Swift
* Which podcasts do we have on GPT-4
* What are the most recent News in the Israel Palestina Conflict

Task: 
For a question, I want to identify all agents that are relevant for this task. Then I want to send the question to each of those agents async, collect the answers, and evaluate if enough information is available to fulfill this task. If not, the agents should try to answer the question again, knowing what they tried before, but not taking care of messages from other agents. This process should be repeated until enough information is collected or a threshold is reached. Then finally, one agent (summarizer) should take all those information to provide a concise answer to the user questions.

**Comment by sonichi** [comment]
[chat_messages](https://microsoft.github.io/autogen/docs/reference/agentchat/conversable_agent/#chat_messages) store messages with each agent separately. For groupchat, each participant talks to the group chat manager directly and the messages sent by other participants are forwarded by the group chat manager. So the message list contains all the messages from each participant. You can filter them by the "name" field though.

**Reply by krlng** [reply]
Okay, so in the group chat situation I should filter the messages I pass into the GroupChatManager.run_chat function or within that function before calling self.send(message...)?

**Reply by ekzhu** [reply]
I think what @sonichi means is that once the group chat session concludes, you can get all the messages from the GroupChatManager's chat_messages.

**Reply by krlng** [reply]
Okay, but during the conversation, aren't all messages sent to GPT all the time? At least when I check _oai_messages of each agent of the groupchat, it looks pretty similar for all agents (with minor differences for those messages, where the agents talked himself)

**Reply by ekzhu** [reply]
Because in group chat, the messages are broadcasted to all agent, but the group chat manager selects agent to speak. So the conversation histories across agents can be quite similar.

**Reply by sonichi** [reply]
We do plan to add a feature to allow selective broadcasting in group chat. @joshkyh plans to do it after #857 is merged.
Depending on your use case, there are a few options:
1. wait for the new feature (or help with accelerating it)
2. use other multi-agent chat that's not group chat
3. customize your group chat

**Reply by krlng** [reply]
Thanks @sonichi . Regarding 2: Are there already any other multi-agent chats that are not group chat? If so, I wasn't able to find them.
